### ISSUE TITLE
Create container from existing container volume

### ISSUE THREAD

**Issue by StevenLiekens** [microsoft/vscode-remote-release#5453]
Issue Type: <b>Feature Request</b>

Hi,

I have an existing container volume (obtained by clicking Clone in Volume), but deleted the actual container. I can still access the volume using 'Explore in Development Container' but what I really need is 'Open in Development Container' to recreate my container from the '.devcontainer' files inside the volume.

Thanks!

Extension version: 0.187.1
VS Code version: Code 1.58.2 (c3f126316369cd610563c75b1b1725e0679adfb3, 2021-07-14T22:10:15.214Z)
OS version: Windows_NT x64 10.0.19043
Restricted Mode: No
Remote OS version: Linux x64 5.4.72-microsoft-standard-WSL2

**Comment by PavelSosin-320** [comment]
This request can hardly be attributed to vscode because this is a Docker feature - volume persistence vs container persistence. See [Volume mount type](https://docs.docker.com/storage/). It has side effect: If it is possible to create a Container from Volume data this container can mount all other volumes in the same Docker engine host, ... recursively. I think this is not always desired.

**Comment by StevenLiekens** [comment]
I'm not sure I follow, is it not possible to achieve what I was hoping for? I don't really see how 'Clone in Volume' is much different from 'Open in Volume' (or however you would describe it).

**Comment by PavelSosin-320** [comment]
@StevenLiekens Yes, it is achievable but causes a certain security risk: it requires running the container in the privileged mode to call docker build that is insecure by itself and will allow to mount all docker volumes on the same host because it will be recursive: create a container, with mounted volume, mount an additional volume, clone project and create a new container, ...
It tends to create the recursive tree of containers-volumes spreading over the entire server/VM filesystem. Ope-in-volume is not recursive because all volumes are mounted by the same container doesn't matter how many mounted volumes the container has. Please, remember that Docker engine has the pre-defined files tree structure with Containers and Volumes subtrees.

**Comment by StevenLiekens** [comment]
But how does 'Clone in Volume' work? I still don't see how creating a new volume and container is more secure than creating a container from an existing volume that was created earlier.

**Comment by PavelSosin-320** [comment]
Clone in Volume does nothing special with Docker Containers, Volumes and doesn't need any privileged mode for the dev container unless DiD mode. To build a new image using devcontainer.json and create a new container from the data stored in the existing volume the container that mounts this volume has to be run in the privileged mode with special capabilities.
The new container that is run with --volume option may mount existing docker volumes bind volumes, i.e. disk directories unless very specific measures unsupported by VSCode will be taken. The safe usage of the requested feature requires certain Docker security proficiency.

**Comment by KalleDK** [comment]
Just to understand the workflow. Incase the devcontainer is deleted, the "correct" way is to also delete the volume and reclone it?

**Comment by octdobre** [comment]
In the _Devvolumes_ section, you can right-click a volume and then click on "Inspect". This will create a container and attach VSCode to it to be inspected. Somehow you **can** create a container from a _Devvolume_ but of course, it is not the desired outcome, the extensions are not installed. 

- Option 1 - delete the volume and reclone
- Option 2 - create a devcontainer.json in a folder somewhere on Desktop or smth and mount the volume

**Comment by StevenLiekens** [comment]
> Somehow you can create a container from a Devvolume but of course, it is not the desired outcome, the extensions are not installed.

I think that gives you a generic container with some basic tools for editing files but not much else. I don't even know what base image it is built from.

Is this not exactly how "Clone in Volume" works...?:

1. create an empty volume for the repository
2. create a generic container and mount the repository volume
3. git clone $repository /workspace && cd /workspace
4. devcontainer build
5. devcontainer open

So what exactly would be the problem if you changed it to use an existing repository volume?

1. Create a generic container and mount the existing volume
2. cd /workspace
3. devcontainer build
4. devcontainer open

So maybe I misunderstand how "Clone in Volume" works, but I don't believe there is any added security risk when you build a dev container from an existing volume vs. pulling the code from the internet before running the devcontainer build. It just doesn't make any sense.

**Comment by octdobre** [comment]
What I meant  @StevenLiekens was that when you go to the Remote-Explorer extension and click on a Devvolume and click on Inspect, it will create a generic container just to inspect and edit the files there.

**Comment by StevenLiekens** [comment]
@octdobre yep I'm completely agreeing with you but I'm also still trying to get information about whether it would be possible to build and connect to a devcontainer from that mode. Someone tried to convince me it isn't feasible but I want a second opinion since MSFT hasn't decided to close this request yet.

**Comment by octdobre** [comment]
@StevenLiekens I have tested one scenario which seems far-fetched but ultimately did not work. 

Here is what I tried:

1. Installed Ubuntu 18 WSL2 Distro in Windows. 
2. Open VSCode with the Remote WSL extension and navigate to any folder in the distro, I chose /home/octobre/testproject as an example
3. Cloned a git repo that also contains a .devcontainer folder
4. Select "Reopen in Container" and it works, the container is created successfully based on the devcontainer.json and mounts the specified volume from the WSL2 instance

Here is what didn't work:

1. I tried to mount the "docker-desktop-data" volume to the WSL2 distro - works
2. Open VSCode with Remote-WSL in Ubuntu distro and navigate to the specified volume in "docker-desktop-data"
 Volumes can be found using explorer.exe at "\\wsl$\docker-desktop-data\version-pack-data\community\docker\volumes"
this also works
3. Reopen in Container, here it fails, even though it worked when I did it from a Ubuntu WSL2 instance.


Here are some of my findings:

- Docker volumes are always saved to "docker-desktop-data" even if you select your WSL2 backend to be Ubuntu.
- Reopen in Container works from Ubuntu WSL2 instance but not from docker-desktop-data

**Comment by octdobre** [comment]
Just in case somebody wants to recover his work from "Clone Repo to Volume" and then delete the container.

Quick recover:
Go to `"\wsl$\docker-desktop-data\version-pack-data\community\docker\volumes"` , navigate inside your desired volume.
Open a Terminal here and commit,push your changes.

Advanced Recover:
Create a folder and inside create a devcontainer definition. You can copy the one from the volume. 
At the end of the definition add this : 
```
"workspaceMount": "source=<insert volume name here>,target=/workspace,type=volume",
"workspaceFolder": "/workspace/<name of volume folder>"
```

Volume name can be found with `docker volume ls`.
Name of folder  can be found in `"\wsl$\docker-desktop-data\version-pack-data\community\docker\volumes"`.
There should be a folder named like the volume name. Inside there should be a _data folder and inside that should be the folder that you are looking for to place in <name of volume folder>.

Ex: Volume name is `myrepo-main-24696f3120bfd7f8fdffdd5aeaf263c3`
Name of the folder is `myrepo`

**Comment by octdobre** [comment]
An interesting feature of the Clone Repo in Volume:
https://www.youtube.com/watch?v=79wq_V7dr84&list=PLbgF63IaYmRAUUVKIZ6n9ipf7t4EF68Rp

**Comment by StevenLiekens** [comment]
> An interesting feature of the Clone Repo in Volume: https://www.youtube.com/watch?v=79wq_V7dr84&list=PLbgF63IaYmRAUUVKIZ6n9ipf7t4EF68Rp

Omg! It's so simple, I never would've thought of it! That's pretty much what I need. Having to enter the repository again is a little weird but only a minor inconvenience. Fantastic find.
